In [43]:
import pandas as pd
import warnings
warnings.filterwarnings("ignore")

- App — название приложения
- Category — категория, которая относится к приложению
- Rating — рейтинг приложения
- Reviews — количество отзывов, оставленных юзерами по приложению
- Size — размер приложения
- Installs — количество установок приложения пользователями
- Type — тип приложения (по денежному критерию)
- Price — цена приложения
- Content Rating — возрастная группа приложения
- Genres — жанр приложения

# Работа с Pandas

1. Загрузите ваш DataFrame в переменную df и выведете на экран первые пять строк 

In [77]:
df = pd.read_csv('googleplaystore.csv')
df.head(5)

App        Category  Rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                                Coloring book moana  ART_AND_DESIGN     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN     4.7   
3                              Sketch - Draw & Paint  ART_AND_DESIGN     4.5   
4              Pixel Draw - Number Art Coloring Book  ART_AND_DESIGN     4.3   

  Reviews  Size     Installs  Type  Price Content Rating  \
0     159   19M      10,000+  Free    0.0       Everyone   
1     967   14M     500,000+  Free    0.0       Everyone   
2   87510  8.7M   5,000,000+  Free    0.0       Everyone   
3  215644   25M  50,000,000+  Free    0.0           Teen   
4     967  2.8M     100,000+  Free    0.0       Everyone   

                      Genres      Last Updated         Current Ver  \
0               Art & Design   January 7, 2018               1.0.0   
1  Art & Design;Pretend Play  January 15, 2018               2.0.0   
2               Art & Design    August 1, 2018               1.2.4   
3               Art & Design      June 8, 2018  Varies with device   
4    Art & Design;Creativity     June 20, 2018                 1.1   

    Android Ver  
0  4.0.3 and up  
1  4.0.3 and up  
2  4.0.3 and up  
3    4.2 and up  
4    4.4 and up

- Приведем названия столбцов в правильный формат

In [45]:
df.columns = [col.lower().replace(' ', '_') for col in df.columns]
df.columns

Index(['app', 'category', 'rating', 'reviews', 'size', 'installs', 'type',
       'price', 'content_rating', 'genres', 'last_updated', 'current_ver',
       'android_ver'],
      dtype='object')

2. Отобразите количество строк и столбцов в DataFrame

In [46]:
df.shape

(10841, 13)

3. Выведете основную информацию о DataFrame методом info(). Проанализируйте есть ли пропуски в столбцах. Напишите вывод о том, какие столбцы содержат пропуски

In [47]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10841 entries, 0 to 10840
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   app             10841 non-null  object 
 1   category        10841 non-null  object 
 2   rating          9367 non-null   float64
 3   reviews         10841 non-null  object 
 4   size            10841 non-null  object 
 5   installs        10841 non-null  object 
 6   type            10840 non-null  object 
 7   price           10841 non-null  float64
 8   content_rating  10840 non-null  object 
 9   genres          10841 non-null  object 
 10  last_updated    10841 non-null  object 
 11  current_ver     10833 non-null  object 
 12  android_ver     10838 non-null  object 
dtypes: float64(2), object(11)
memory usage: 1.1+ MB


1. У нас есть 10841 строка и 13 столбцов;
2. `rating` и `price` имеют тип данных `float64`. Остальные имеют тип данных `object` (11 штук);
3. Столбцы `rating`, `type`, `content_rating`, `current_ver` и `android_ver` содержат нули.

4. Удалите или замените пропуски известными вам методами. (Подсказка: `dropna()` или `fillna()`, выбор того или иного метода должен быть аргументирован).

In [48]:
df.describe().round(2)

rating     price
count  9367.00  10841.00
mean      4.19      1.03
std       0.54     15.95
min       1.00      0.00
25%       4.00      0.00
50%       4.30      0.00
75%       4.50      0.00
max      19.00    400.00

In [50]:
df['rating'].fillna(df['rating'].median(), inplace=True)

In [51]:
df.describe(include='object')

app category reviews                size    installs   type  \
count    10841    10841   10841               10841       10841  10840   
unique    9660       34    6002                 462          22      3   
top     ROBLOX   FAMILY       0  Varies with device  1,000,000+   Free   
freq         9     1972     596                1695        1579  10039   

       content_rating genres    last_updated         current_ver android_ver  
count           10840  10841           10841               10833       10838  
unique              6    120            1378                2832          33  
top          Everyone  Tools  August 3, 2018  Varies with device  4.1 and up  
freq             8714    842             326                1459        2451

In [52]:
df.dropna(inplace=True)

5. Проверьте нет ли пропусков после произведенной обработки

In [53]:
df.shape

(10829, 13)

In [54]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 10829 entries, 0 to 10840
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   app             10829 non-null  object 
 1   category        10829 non-null  object 
 2   rating          10829 non-null  float64
 3   reviews         10829 non-null  object 
 4   size            10829 non-null  object 
 5   installs        10829 non-null  object 
 6   type            10829 non-null  object 
 7   price           10829 non-null  float64
 8   content_rating  10829 non-null  object 
 9   genres          10829 non-null  object 
 10  last_updated    10829 non-null  object 
 11  current_ver     10829 non-null  object 
 12  android_ver     10829 non-null  object 
dtypes: float64(2), object(11)
memory usage: 1.2+ MB


### Вывод:
Сейчас количество ненулевых значений по каждому столбцу равно размеру датафрейма, а значит нулевых значений неты

6. Выведите на экран минимум и максимум из столбца `Price`:

In [55]:
print(f'Минимум столбца `price` - {df["price"].min()}\n\
Максимум столбца `price` - {df["price"].max()}')

Минимум столбца `price` - 0.0
Максимум столбца `price` - 400.0


7. Выведите на экран медиану и среднее арифметическое столбцов:
 - Rating
 - Reviews

- Столбец `reviews` имеет тип данных `object`, хотя по факту содержит тип данных `int`. Изменим тип и посчитаем

In [56]:
df['reviews'].unique()

array(['159', '967', '87510', ..., '603', '1195', '398307'], dtype=object)

In [57]:
df['reviews'] = df['reviews'].astype(int)

In [58]:
df['reviews'].dtype

dtype('int64')

In [59]:
print(f'Медиана столбца `rating` - {df["rating"].median()}\n\
Среднее значение столбца `rating` - {df["rating"].mean().round(2)}')
print(f'Медиана столбца `reviews` - {df["reviews"].median()}\n\
Среднее значение столбца `reviews` - {df["reviews"].mean().round(2)}')

Медиана столбца `rating` - 4.3
Среднее значение столбца `rating` - 4.21
Медиана столбца `reviews` - 2100.0
Среднее значение столбца `reviews` - 444601.77


8. Выведите на экран все уникальные значения категориального столбца Genres

In [60]:
df['genres'].unique()

array(['Art & Design', 'Art & Design;Pretend Play',
       'Art & Design;Creativity', 'Art & Design;Action & Adventure',
       'Auto & Vehicles', 'Beauty', 'Books & Reference', 'Business',
       'Comics', 'Comics;Creativity', 'Communication', 'Dating',
       'Education;Education', 'Education', 'Education;Creativity',
       'Education;Music & Video', 'Education;Action & Adventure',
       'Education;Pretend Play', 'Education;Brain Games', 'Entertainment',
       'Entertainment;Music & Video', 'Entertainment;Brain Games',
       'Entertainment;Creativity', 'Events', 'Finance', 'Food & Drink',
       'Health & Fitness', 'House & Home', 'Libraries & Demo',
       'Lifestyle', 'Lifestyle;Pretend Play',
       'Adventure;Action & Adventure', 'Arcade', 'Casual', 'Card',
       'Casual;Pretend Play', 'Action', 'Strategy', 'Puzzle', 'Sports',
       'Music', 'Word', 'Racing', 'Casual;Creativity',
       'Casual;Action & Adventure', 'Simulation', 'Adventure', 'Board',
       'Trivia', 'Role 

9. Сгруппируйте данные по столбцу `Genres` и посчитайте для каждого жанра средний и медианный рейтинг. Представьте результат в виде нового DataFrame,который будет сохранен в переменную `grouped_df`, где одна колонка будет содержать названия жарнов, а две другие буду содержать в себе средний и медианный рейтинги. Напишите краткий вывод.

In [66]:
grouped_df = df.groupby(by='genres', as_index=True).agg(mean_rating = ('rating', 'mean'),
                            median_rating = ('rating', 'median'))
grouped_df.shape

(119, 2)

10. Из получившегося датафрейма выведите на экран только те жанры, чей медианный рейтинг больше 4.5. Напишите краткий вывод.

In [68]:
grouped_df.loc[grouped_df['median_rating'] > 4.5]

mean_rating  median_rating
genres                                                 
Adventure;Brain Games           4.600000            4.6
Art & Design;Creativity         4.400000            4.7
Board;Pretend Play              4.800000            4.8
Comics;Creativity               4.800000            4.8
Entertainment;Creativity        4.533333            4.6
Health & Fitness;Education      4.700000            4.7
Music;Music & Video             4.533333            4.6
Puzzle;Education                4.600000            4.6
Strategy;Action & Adventure     4.600000            4.6

### Вывод:
- Примерно 7% от всего рейтинга составляют жанры, имеющие рейтинг выше 4.5

11. Выведете кол-во дубликатов в основном DataFrame (df) и удалите их 

In [71]:
df.duplicated().sum()

np.int64(483)

In [72]:
df.drop_duplicates(inplace=True)

In [73]:
df.duplicated().sum()

np.int64(0)

12. Посчитатайте какое количество приложений содержится в каждом жанре (value_counts).В получившемся Series выведите на экран 10 строк с наибольшим количеством приложений отсортированных по убыванию.

In [76]:
df['genres'].value_counts().head(10)

genres
Tools              840
Entertainment      587
Education          526
Business           427
Medical            408
Productivity       407
Personalization    386
Lifestyle          372
Communication      366
Sports             364
Name: count, dtype: int64